jsapp/js/account/plans/plan.component.tsx

```diff
@@ -249,93 +292,76 @@ export default function Plan() {
-      state.subscribedProduct.filter(
-        (subscription: BaseSubscription) =>
-          subscription.items[0].price.product.id === productId
-      ),
-    [state.subscribedProduct]
-  );
+  const getSubscribedProduct = useCallback(getSubscriptionsForProductId, []);
 
   const isSubscribedProduct = useCallback(
     (product: Price) => {
-      if (!product.prices.unit_amount && !hasActiveSubscription) {
+      if (!product.prices?.unit_amount && !hasActiveSubscription) {
         return true;
       }
 
-      const subscriptions = getSubscriptionsForProductId(product.id);
+      const subscriptions = getSubscribedProduct(
+        product.id,
+        state.subscribedProduct
+      );
 
-      if (subscriptions.length > 0) {
+      if (subscriptions && subscriptions.length > 0) {
         return subscriptions.some(
-          (subscription: BaseSubscription) =>
+          (subscription: SubscriptionInfo) =>
             subscription.items[0].price.id === product.prices.id &&
             hasManageableStatus(subscription)
         );
       }
       return false;
     },
-    [state.subscribedProduct, state.intervalFilter]
+    [state.subscribedProduct, state.intervalFilter, state.products]
   );
 
-  const shouldShowManage = useCallback(
-    (product: Price) => {
-      const subscriptions = getSubscriptionsForProductId(product.id);
-      if (!subscriptions.length || !state.organization?.id) {
-        return false;
-      }
-
-      return subscriptions.some((subscription: BaseSubscription) =>
-        hasManageableStatus(subscription)
-      );
-    },
-    [state.subscribedProduct]
-  );
-
-  const upgradePlan = (priceId: string) => {
-    if (!priceId || areButtonsDisabled) {
-      return;
-    }
-    setAreButtonsDisabled(true);
-    postCheckout(priceId, state.organization?.id)
-      .then((data) => {
-        if (!data.url) {
-          notify.error(t('There has been an issue, please try again later.'));
-        } else {
-          window.location.assign(data.url);
-        }
-      })
-      .catch(() => setAreButtonsDisabled(false));
+  const dismissConfirmModal = () => {
+    setConfirmModal((prevState) => {
+      return {...prevState, price: null, subscription: null};
+    });
+    setIsBusy(false);
   };
 
-  const managePlan = () => {
-    if (!state.organization?.id || areButtonsDisabled) {
+  const buySubscription = (price: BasePrice) => {
+    if (!price.id || isBusy || !state.organization?.id) {
       return;
     }
-    setAreButtonsDisabled(true);
-    postCustomerPortal(state.organization.id)
-      .then((data) => {
-        if (!data.url) {
-          notify.error(t('There has been an issue, please try again later.'));
-        } else {
-          window.location.assign(data.url);
-        }
-      })
-      .catch(() => setAreButtonsDisabled(false));
+    setIsBusy(true);
+    if (activeSubscriptions.length) {
+      if (
+        activeSubscriptions[0].items[0].price.unit_amount < price.unit_amount
```

Comment:
Do you think this long chaing will not bit us in future in some edge case where some array is empty?